### ISSUE TITLE
Array of enums whose literals contain special characters cannot be bound in PostgreSQL

### ISSUE THREAD

**Issue by tbroyer** [jOOQ/jOOQ#8342]
### Expected behavior and actual behavior:

I have a table with columns as arrays of enums:
```sql
CREATE TYPE access_right AS ENUM('foo-bar', 'baz-qux');
-- actual enum has many more values

CREATE TABLE user (
  id_user TEXT NOT NULL,
  read    access_right[] NOT NULL DEFAULT '{}',
  write   access_right[] NOT NULL DEFAULT '{}',
  PRIMARY KEY (id_user),
  CHECK (read @> write)
);
```

I'm using jOOQ Codegen to generate Java code, which generates something like:

<details>
<summary>User.java</summary>

```java
public class User extends TableImpl<Record> {
  // …
  public final TableField<Record, AccessRight[]> READ =
      createField(
          "read",
          org.jooq.impl.SQLDataType.VARCHAR
              .asEnumDataType(AccessRight.class)
              .getArrayDataType(),
          this,
          "");

  /** The column <code>siar.utilisateur.gestion</code>. */
  public final TableField<Record, AccessRight[]> write =
      createField(
          "write",
          org.jooq.impl.SQLDataType.VARCHAR
              .asEnumDataType(AccessRight.class)
              .getArrayDataType(),
          this,
          "");

  // …
}
```

</details>
<details>
<summary>AccessRight.java</summary>

```java
public enum AccessRight implements EnumType {
  foo_bar("foo-bar"),
  baz_qux("baz-qux");

  private final String literal;

  private AccessRight(String literal) {
    this.literal = literal;
  }

  /** {@inheritDoc} */
  @Override
  public Catalog getCatalog() {
    return getSchema() == null ? null : getSchema().getCatalog();
  }

  /** {@inheritDoc} */
  @Override
  public Schema getSchema() {
    return /* my schema */;
  }

  /** {@inheritDoc} */
  @Override
  public String getName() {
    return "droit";
  }

  /** {@inheritDoc} */
  @Override
  public String getLiteral() {
    return literal;
  }
```

</details>

An update will incorrectly use `"foo_bar"` and `"baz_qux"` as enum values instead of `"foo-bar"` and `"baz-qux"` (haven't tried inserts yet).

This looks to me as a regression of #3214, reintroduced in https://github.com/jOOQ/jOOQ/commit/97a58e9d72cdf0522a4147fc8246ff0405f71b9c when fixing #4162 one year later: there's an `if (family == POSTGRES …` check in the `else` block following the introduced `if (family == POSTGRES)`; that block should have been merged into the introduced `if (family == POSTGRES)` section.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

```java
String id = …;
AccessRight[] read = …;
AccessRight[] write = …;
User user = dsl
        .update(USER)
        .set(USER.READ, read)
        .set(USER.WRITE, write)
        .where(USER.ID_USER.eq(id))
        .returning()
        .fetchOne()
        .into(User.class);
```

This fails with:
```
org.jooq.exception.DataAccessException: SQL [update "c"."user" set "read" = ?::"c"."access_right"[], "write" = ?::"c"."access_right"[] where "c"."user"."id_user" = ? returning "c"."user"."id_user", "c"."user"."read", "c"."user"."write"]; ERROR: invalid input value for enum access_right: "foo_bar"
[…]
Caused by: org.postgresql.util.PSQLException: ERROR: invalid input value for enum access_right: "foo_bar"
```

If I print the SQL with inline parameters (using `.getSQL(ParamType.INLINE)`), it gives me:
```sql
update "c"."user" set "read" = cast('{"foo_bar"}' as access_right[]), "write" = cast('{"foo_bar"}' as access_right[]) where "c"."user"."id_user" = 'foo' returning "c"."user"."id_user", "c"."user"."read", "c"."user"."write"
```

### Versions:

- jOOQ: 3.11.9
- Java: OpenJDK 11.0.2
- Database (include vendor): Postgres 11.1 (from `postgres:11` Docker image)
- OS: Arch Linux
- JDBC Driver (include name if inofficial driver): `org.postgresql:postgresql:42.2.5`

**Comment by tbroyer** [comment]
MCVE at https://github.com/atolcd-contrib/jOOQ-mcve/tree/issue-8342

**Comment by lukaseder** [comment]
Thanks a lot for your report and the MCVE. With a few tweaks, I could reproduce the issue. I'm not convinced this is a regression, it may well be that this just never worked for enum arrays in PostgreSQL, whose `EnumType.getLiteral()` differs from the `Enum.name()`. In particular, the `org.jooq.util.postgres.PostgresUtils.toPGString(Object o)` method is not capable of correctly handling `EnumType` right now.

This should be an easy fix.

**Comment by lukaseder** [comment]
This issue affects both `val()` and `inline()` versions of generated enum arrays. The fix fixes both.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.12. Backport scheduled for 3.11.10 (#8379) and 3.10. 9 (#8380)

**Comment by tbroyer** [comment]
Ah right, my diagnostic was probably wrong; still, those lines will never be reached in Postgres, so might just as well be deleted (or at least the Postgres case removed from the test): https://github.com/jOOQ/jOOQ/blob/8aa087e5fb8aa31dc4d78b92b07594fcf648cf86/jOOQ/src/main/java/org/jooq/impl/DefaultBinding.java#L983-L985

Thanks for the fix; eagerly awaiting 3.11.10 now :wink:

**Comment by lukaseder** [comment]
Hmm, interesting. I'll double check: You're right about the code not being executed at that location. I think that code was about cross schema enum casts being necessary. Will re-check on monday, and possibly fix another issue.
